test/server/guarddog_impl_test.cc

```diff
@@ -355,6 +438,360 @@ TEST_P(GuardDogTestBase, AtomicIsAtomicTest) {
+            "typed_config": {
+              "@type": "type.googleapis.com/google.protobuf.Empty"
+            }
+          },
+          "event": "KILL"
+        }
+      )EOF",
+        R"EOF(
+        {
+          "config": {
+            "name": "TestAction",
+            "typed_config": {
+              "@type": "type.googleapis.com/google.protobuf.Empty"
+            }
+          },
+          "event": "MEGAMISS"
+        }
+      )EOF",
+        R"EOF(
+        {
+          "config": {
+            "name": "TestAction",
+            "typed_config": {
+              "@type": "type.googleapis.com/google.protobuf.Empty"
+            }
+          },
+          "event": "MISS"
+        }
+      )EOF"};
+  }
+
+  void SetupFirstDog(const NiceMock<Configuration::MockMain>& config) {
+    initGuardDog(fake_stats_, config);
+    first_dog_ = guard_dog_->createWatchDog(api_->threadFactory().currentThreadId(), "test_thread");
+    guard_dog_->forceCheckForTest();
+  }
+
+  std::vector<std::string> actions_;
+  std::vector<std::string> events_;
+  TestGuardDogActionFactory<Envoy::ProtobufWkt::Empty> factory_;
+  Registry::InjectFactory<Configuration::GuardDogActionFactory> register_factory_;
+  NiceMock<Stats::MockStore> fake_stats_;
+  WatchDogSharedPtr first_dog_;
+  WatchDogSharedPtr second_dog_;
+};
+
+INSTANTIATE_TEST_SUITE_P(TimeSystemType, GuardDogActionsTest,
+                         testing::ValuesIn({TimeSystemType::Real, TimeSystemType::Simulated}));
+
+TEST_P(GuardDogActionsTest, MissShouldOnlyReportRelevantThreads) {
+  const NiceMock<Configuration::MockMain> config(100, DISABLE_MEGAMISS, DISABLE_KILL,
+                                                 DISABLE_MULTIKILL, 0, getActionsConfig());
+  SetupFirstDog(config);
+  second_dog_ = guard_dog_->createWatchDog(api_->threadFactory().currentThreadId(), "test_thread");
+  time_system_->advanceTimeWait(std::chrono::milliseconds(99));
+  second_dog_->touch();
+  ASSERT_EQ(events_.size(), 0);
+
+  time_system_->advanceTimeWait(std::chrono::milliseconds(2));
+  guard_dog_->forceCheckForTest();
+
+  EXPECT_THAT(events_, ElementsAre(absl::StrCat(
+                           "MISS : ", api_->threadFactory().currentThreadId().debugString())));
+}
+
+TEST_P(GuardDogActionsTest, MissShouldBeAbleToReportMultipleThreads) {
+  const NiceMock<Configuration::MockMain> config(100, DISABLE_MEGAMISS, DISABLE_KILL,
+                                                 DISABLE_MULTIKILL, 0, getActionsConfig());
+  initGuardDog(fake_stats_, config);
+  first_dog_ = guard_dog_->createWatchDog(api_->threadFactory().currentThreadId(), "test_thread");
+  second_dog_ = guard_dog_->createWatchDog(api_->threadFactory().currentThreadId(), "test_thread");
+
+  first_dog_->touch();
+  second_dog_->touch();
+
+  ASSERT_EQ(events_.size(), 0);
+
+  time_system_->advanceTimeWait(std::chrono::milliseconds(101));
+  guard_dog_->forceCheckForTest();
+  EXPECT_THAT(events_, ElementsAre(absl::StrCat(
```

Comment:
Wondering if use of gmock for watchdog actions would make the flow in this test more natural.  I can see the counter argument of gmock requiring a custom createGuardDogActionFromProto implementation.

Reply:
It seems a mock would have less fidelity, these test examples could also provide documentation / examples of extending the `GuardDogAction` class, as well as the point you brought up.

The aim of this test is to check that the guarddog can report multiple threads that have a miss. Could you clarify on what you mean by 'more natural'? Thanks